pkg/discovery/clustersetip/clustersetip.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error reading ClustersetIPClusterSize")
+	}
+
+	err = json.Unmarshal([]byte(configMap.Data[clustersetIPCidrRange]), &clustersetIPInfo.CIDR)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error reading ClustersetIPCidrRange")
+	}
+
+	clustersetIPInfo.Clusters, err = cidr.ExtractClusterInfo(configMap)
+
+	return &clustersetIPInfo, configMap, err //nolint:wrapcheck // No need to wrap
+}
+
+func AssignClustersetIPs(clustersetIPInfo *Info, netconfig Config, status reporter.Interface) (string, error) {
+	status.Start("Assigning ClustersetIP IPs")
+	defer status.End()
+
+	clustersetIPCIDR := netconfig.ClustersetIPCIDR
+	clusterID := netconfig.ClusterID
+	var err error
+
+	if clustersetIPCIDR == "" {
+		// ClustersetIPCIDR not specified by the user
+		if isCIDRPreConfigured(clusterID, clustersetIPInfo.Clusters) {
+			// clustersetipCidr already configured on this cluster
+			clustersetIPCIDR = clustersetIPInfo.Clusters[clusterID].CIDRs[0]
+			status.Success("Using pre-configured clustersetip CIDR %s", clustersetIPCIDR)
+		} else {
+			// no clustersetipCidr configured on this cluster
+			clustersetIPCIDR, err = AllocateClustersetIPCIDR(clustersetIPInfo)
+			if err != nil {
+				return "", status.Error(err, "unable to allocate clustersetip CIDR")
+			}
+
+			status.Success(fmt.Sprintf("Allocated clustersetip CIDR %s", clustersetIPCIDR))
+		}
+	} else {
+		// ClustersetIP enabled, clustersetIPCIDR specified by user
+		if isCIDRPreConfigured(clusterID, clustersetIPInfo.Clusters) {
+			// clustersetipCidr pre-configured on this cluster
+			clustersetIPCIDR = clustersetIPInfo.Clusters[clusterID].CIDRs[0]
+			status.Warning("A pre-configured clustersetip CIDR %s was detected - not using the specified CIDR %s",
+				clustersetIPCIDR, netconfig.ClustersetIPCIDR)
+		} else {
+			// clustersetipCidr as specified by the user
+			err := cidr.CheckForOverlappingCIDRs(clustersetIPInfo.Clusters, netconfig.ClustersetIPCIDR, netconfig.ClusterID)
+			if err != nil {
+				return "", status.Error(err, "error validating overlapping clustersetip CIDRs %s", clustersetIPCIDR)
+			}
+
+			status.Success("Using specified clustersetip CIDR %s", clustersetIPCIDR)
+		}
+	}
+
+	return clustersetIPCIDR, nil
+}
+
+func ValidateExistingClustersetIPNetworks(ctx context.Context, client controllerClient.Client, namespace string) error {
+	clustersetIPInfo, _, err := GetClustersetIPNetworks(ctx, client, namespace)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return errors.Wrap(err, "error getting existing clustersetip configmap")
+	}
+
+	if clustersetIPInfo != nil {
+		if err = cidr.IsValid(clustersetIPInfo.CIDR); err != nil {
+			return errors.Wrap(err, "invalid ClustersetIPCidrRange")
+		}
+	}
+
+	return nil
+}
+
+func AllocateAndUpdateClustersetIPCIDRConfigMap(ctx context.Context, brokerAdminClient controllerClient.Client, brokerNamespace string,
```

Comment:
This is quite a mouthful 😄 How about:

```suggestion
func AllocateCIDRFromConfigMap(ctx context.Context, brokerAdminClient controllerClient.Client, brokerNamespace string,
```

Reply:
Thanks, done.